lightningd/peer_control.c

```diff
@@ -1611,6 +1613,75 @@ static const struct json_command listpeers_command = {
 /* Comment added to satisfice AUTODATA */
 AUTODATA(json_command, &listpeers_command);
 
+
+static void json_add_peerchannels(struct lightningd *ld,
+                          struct json_stream *response,
+                          struct peer *peer,
+			  enum channel_state *state,
+                          const enum log_level *ll)
+{
+	struct channel *channel;
+
+        if (!state)
+	        json_add_uncommitted_channel(response, peer->uncommitted_channel);
```

Comment:
I'm sure that I'm missing some things here, but I didn't catch this part here of your code. 

Why if the state is unspecified in the command you are returning only the uncommitted channels? The state here is useful to get the channel with an ON_CHAIN (closed state). 

I think it is rare to filter the channels by a status here like "CHANNELD_NORMAL", I usually preferer with API to simplify the life of the user. I would be happy to have only a couple of status propriety like "closed", "normal", "pending". So in this way, you can have that the following state correspond to the 

- closed -> (ONCHAIN, CLOSED)
- normal -> (CHANNELD_NORMAL)
- pending -> (CHANNELD_SHUTTING_DOWN, CLOSINGD_SIGEXCHANGE, CLOSINGD_COMPLETE)

I think this makes cooler the API of the status, now I know that CLOSINGD_SIGEXCHANGE can be a pending status of the channel, this means that some operations need to happen.

p.s: I think this is a good moment to query the DB to retrieval the closed channels with the status "closed"

Reply:
Ok, please allow me to explain the case with unspecified state. The original listpeers code did not have a state filter, and it returned uncommitted channels AND all the other relevant channels, and I simply retained that logic here, with the exception that if the state filter is specified, then uncommitted channels are omitted from the result. To be clear, if no state is specified, the result will contain uncommitted channels AND all other channels just like the original behavior of listpeers.

I think the DB retrieval of closed channels should be kept separate from this issue, to be handled by the plugin I am working on. Perhaps not to muddy the waters, we should even take out the filter option from this PR altogether.

Reply:
>Ok, please allow me to explain the case with unspecified state. The original listpeers code did not have a state filter, and it returned uncommitted channels AND all the other relevant channels, and I simply retained that logic here, with the exception that if the state filter is specified, then uncommitted channels are omitted from the result. To be clear, if no state is specified, the result will contain uncommitted channels AND all other channels just like the original behavior of listpeers.

Usually in the c-lightning commands when you don't specify the status the command line returns all the data without filter by status. If the status is specified, there is filtering with the status applied to it.

>I think the DB retrieval of closed channels should be kept separate from this issue, to be handled by the plugin I am working on. Perhaps not to muddy the waters, we should even take out the filter option from this PR altogether.

I think that if you specify the status here, you need to add also the query to the closed channels, otherwise, if you query by CLOSED status you have no result, but this in a node that works for a long time can be wrong (and also confusing for the end user?).